### ISSUE TITLE
Reimplement question

### ISSUE THREAD

**Issue by MyYaYa** [FengZhenhua/Wing-Loss#2]
Hi, Feng,

Thanks for your excellent work on Wing Loss.
I'm reimplementing the simple cnn6 networks. Aftering reading your paper, I do training on the aflw-full-protocol dataset, my steps is below.
1. PDB, with procrustes analysis and PCA, I duplicate the fewer images (make every bin is equal, but for those bins that have too fewer images, I set the upper bound--20 dumplicates)
2. Data Augmentation, random flip, random rotate(-30, 30), random little box jitter, random little image brightness changes and so on
3. CNN6 network. Firstly, I use batch normalization (why I use bn is explained latter). input image is normalize to (-1, 1), and output landmark of the network is normalize to (0~1)
4, hyperparameter. learning_rate: 3e-5 to 3e-7 with cosine warm restart(SGDR), weight_decay:5e-4, using SGDM with momentum 0.9

Under these setting, the training loss and training nme is well, but on the validation dataset, the nme is not good, even poor.
I guess it's overfitting, but I think that data is enough, and the CNN6 is simple enough, how this overfitting circumstance happens?
![image](https://user-images.githubusercontent.com/6195920/44968204-2b782d80-af79-11e8-9dbd-e7c5fb42757a.png)
As shown in figure, training_nme_per_batch is well, but eval_nme is up and down around 0.10.

Here are what I want to know:
1. How do you process aflw-full-protocol dataset properly? 
2. I see the model mat file, Is the output of network is arange from 0 to 64? not normalize to 0~1?
3. I also try not to use bn, but the performance is even worse.

**Comment by FengZhenhua** [comment]
Hi,

For you implementation:

It sounds that your PDB and data augmentation are correct.

For the CNN-6 network, I did not use BN because it performs worse that the one without BN.

I resize the intensity of all the pixels of an input image to (0\~1) rather than (-1\~1). But I do not think this is a big difference. 

For the output of the landmarks, I did not resize them to (0\~1). They are between 1 and 64. If you resize them to (0~1), you should increase the learning rate. Maybe you should increase it from 3e-5\~3e-7 to 1e-3\~1e-5.

For your question "How do you process aflw-full-protocol dataset properly?", I do not quite understand what does it mean. Could you please you explain it further?

Hope my answer is helpful for you.

Zhenhua

**Comment by MyYaYa** [comment]
@FengZhenhua  Thanks!
The question, "How do you process aflw dataset properly?", means how do you do PDB and Data Augmentation.
You have told me that what I did is correct, so this question is answered nicely.

And fortunately, I found the key reason of my reimplementation's worse performance. It's the random flip. I use TF's Dataset Class for feeding data, and also use Dataset's inner func for augmentation. Though I check the ouput of my Dataset, that is correct, the training will be like the picture I have shown. Now, I cancel the random flip, the performance is back to normal.

When I check out the actual reason of degradation caused by random flip, I will explain it here.

And lastly, can you tell me the insight of increasing learning rate for smaller range of output?

**Comment by MyYaYa** [comment]
I Found the reason why random flip get worse.
It's the order of landmarks. It means that the order of left eyes represent the right eyes.

It's so stupid, sorry for bothering.

**Comment by FengZhenhua** [comment]
@MyYaYa  Great to hear that you have solved your problem. For the learning rate, I think it would be easy to explain. If you use the output between 0 and 100, it amplifies the gradient by 100 times than the one using value between 0 and 1. Hence, the learning rate should be increased when you use smaller range of outputs.

A further reminder for image flip is that, you also have to carefully check flipped landmarks not only in order but also in locations.

**Comment by MyYaYa** [comment]
@FengZhenhua Thanks, I will take care of it. Now I close the issue.

**Comment by protossw512** [comment]
@MyYaYa Can I ask what is your test NME on 300W dataset?
